packages/runtime/container-runtime/src/summary/runningSummarizer.ts

```diff
@@ -556,6 +567,88 @@ export class RunningSummarizer implements IDisposable {
 		return resultsBuilder.build();
 	}
 
+	/**
+	 * Tries to summarize with retries where retry is based on the failure params.
+	 * For example, summarization may be retried for failures with "retryAfterSeconds" param.
+	 */
+	private async trySummarizeWithRetries(
+		reason: SummarizeReason,
+		cancellationToken: ISummaryCancellationToken,
+	) {
+		let maxAttempts = defaultMaxAttempts;
+		let currentAttempts = 0;
+
+		let done = false;
+		do {
+			if (this.cancellationToken.cancelled) {
+				return;
+			}
+
+			currentAttempts++;
+			const summarizeOptions: ISummarizeOptions = {
+				refreshLatestAck: false,
+				fullTree: false,
+			};
+			const summarizeProps: ISummarizeTelemetryProperties = {
+				reason,
+				summaryAttempts: currentAttempts,
+				...summarizeOptions,
+			};
+			const summarizeResult = this.generator.summarize(
+				summarizeProps,
+				summarizeOptions,
+				cancellationToken,
+			);
+
+			const ackNackResult = await summarizeResult.receivedSummaryAckOrNack;
+			if (ackNackResult.success) {
+				return;
+			}
+
+			const submitSummaryResult = await summarizeResult.summarySubmitted;
+			let retryAfterSeconds: number | undefined;
+
+			// Update max attempts and retry params from the failure result.
+			// If submit summary failed, use the params from "summarySubmitted" result. Else, use the params
+			// from "receivedSummaryAckOrNack" result.
+			// Note: Check "summarySubmitted" result first because if it fails, ack nack would fail as well.
+			if (!submitSummaryResult.success) {
+				// The max attempts for submit failures can be overridden via a feature flag. This allows us to
+				// tweak this as per telemetry data until we arrive at a stable number.
+				// If its set to a number higher than `maxAttemptsForSubmitFailures`, it will be ignored.
+				const overrideMaxAttempts = this.mc.config.getNumber(
+					"Fluid.Summarizer.AttemptsForSubmitFailures",
+				);
+				maxAttempts =
+					overrideMaxAttempts && overrideMaxAttempts <= maxAttemptsForSubmitFailures
+						? overrideMaxAttempts
+						: maxAttemptsForSubmitFailures;
+				retryAfterSeconds = submitSummaryResult.data?.retryAfterSeconds;
+			} else {
+				maxAttempts = defaultMaxAttempts;
+				retryAfterSeconds = ackNackResult.data?.retryAfterSeconds;
+			}
+
+			// If "retryAfterSeconds" is part of failure params and we haven't done max attempts, add a delay
+			// and retry another attempt.
+			if (retryAfterSeconds !== undefined && currentAttempts < maxAttempts) {
+				this.mc.logger.sendPerformanceEvent({
+					eventName: "SummarizeAttemptDelay",
+					duration: retryAfterSeconds,
+					summaryNackDelay: ackNackResult.data?.retryAfterSeconds !== undefined,
+					stage: submitSummaryResult.data?.stage,
+					dynamicRetries: true, // To differentiate this telemetry from regular retry logic
+					...summarizeProps,
+				});
+				await delay(retryAfterSeconds * 1000);
+			} else {
+				done = true;
```

Comment:
style idea: I think if you pull `this.stopSummarizerCallback` out of this function you can ditch the `done` variable and just do an early return here if retryAfter is unset or you're out of attempts